File: src/logic/evaluation.js
const { evaluateConditions } = require("./conditions");
const { getMetrics } = require("./metrics");
const { evaluateOutput } = require("./outputs");
const { getResultData } = require("./results");
const { checkSettings } = require("./settings");

var settings = []
var variables = []

async function evaluateSRE(tree, contextVariables = {}, filterUsers) {
    let results = [];
    settings = tree.settings;
    // No outputs
    let no_outputs = settings.find(element => element.key == "no_outputs");
    let treeMetrics = tree.metrics;
    variables = JSON.parse(contextVariables) || {};

    if (filterUsers.length > 0) {
        await Promise.all(
            filterUsers.map(async (userInfo, index) => {
                // Globbbal data/Context
                let userData = userInfo._doc;
                let result = await evaluateNodes(tree.nodes, userData);
                let user = userInfo.msisdn;
                let resultOutputs = result.outputs;
                let metrics = await getMetrics(treeMetrics, userData);
                if (resultOutputs.length > 0) {
                    let outputs = await checkSettings(resultOutputs, settings);
                    results.push({ user, outputs, metrics });
                } else {
                    if (no_outputs && no_outputs.value == true) {
                        let outputs = [];
                        results.push({ user, outputs, metrics });
                    }
                }

            })
        );
    }

    //suffle results
    results.sort(() => Math.random() - 0.5);

    let dataResult = await getResultData(results, tree.name);

    return dataResult;
}

async function evaluateSRESF(tree, contextVariables = {}, filterUsers) {
    let results = [];
    settings = tree.settings;
    // No outputs
    let no_outputs = settings.find(element => element.key == "no_outputs");
    let treeMetrics = tree.metrics;
    variables = JSON.parse(contextVariables) || {};

    if (filterUsers.length > 0) {
        await Promise.all(
            filterUsers.map(async (userInfo, index) => {
                // Globbbal data/Context
                let userData = userInfo
                let result = await evaluateNodes(tree.nodes, userData);
                let user = userInfo.msisdn;
                let resultOutputs = result.outputs;
                let metrics = await getMetrics(treeMetrics, userData);
                if (resultOutputs.length > 0) {
                    let outputs = await checkSettings(resultOutputs, settings);
                    results.push({ user, outputs, metrics });
                } else {
                    if (no_outputs && no_outputs.value == true) {
                        let outputs = [];
                        results.push({ user, outputs, metrics });
                    }
                }

            })
        );
    }

    //suffle results
    results.sort(() => Math.random() - 0.5);

    let dataResult = await getResultData(results);

    return dataResult;
}

async function evaluateNodes(nodes, userData, exclusive = false) {
    let results = {
        paths: [],
        outputs: []
    }

    if (nodes && nodes.length == 0) {
        return results;
    }

    for (let node of nodes) {
        // Skip disabled nodes
        if (node.enable == false) { continue; }

        // Evaluate node
        let nodeResult = {
            paths: [],
            outputs: [],
        }
        if (evaluateConditions(node.conditions, userData, variables)) {
            nodeResult.paths.push(node.description);

            if (node.output) {
                let outputResult = await evaluateOutput(node.outputs, userData);
                nodeResult.outputs = [...nodeResult.outputs, ...outputResult];
            }
            if (node.nodes && node.nodes.length > 0) {
                let childResult = await evaluateNodes(node.nodes, userData, node.exclusive);
                nodeResult.paths = [...nodeResult.paths, ...childResult.paths];
                nodeResult.outputs = [...nodeResult.outputs, ...childResult.outputs];
            }

            results.paths.push(nodeResult.paths);
            results.outputs = [...results.outputs, ...nodeResult.outputs];

            if (exclusive) {
                break;
            }
        }
    }

    return results;
}




module.exports = { evaluateSRE, evaluateSRESF };

File: py_logic/python_lambdas/segmentation-engine-core-python-evaluation/evaluation.py
from conditions import *
from metrics import *
from outputs import *
from settings import *
from results import *
import random
def evaluate_sresf(tree, context_variables={}, filter_users=[]):
    print("start evaluation")
    results = []
    settings = tree['settings']
    no_outputs = next((element for element in settings if element['key'] == 'no_outputs'), None)
    tree_metrics = tree['metrics']
    variables = json.loads(context_variables) if context_variables else {}
    data_result = []
    if filter_users:
        print(len(filter_users))
        for index, user_info in enumerate(filter_users):
            user_data = user_info
            result = evaluate_nodes(tree['nodes'], user_data)
            user = user_info['msisdn']
            result_outputs = result['outputs']
            metrics = get_metrics(tree_metrics, user_data)

            if result_outputs:
                outputs = check_settings(result_outputs, settings)
                results.append({'user': user, 'outputs': outputs, 'metrics': metrics})
            elif no_outputs and no_outputs['value']:
                outputs = []
                results.append({'user': user, 'outputs': outputs, 'metrics': metrics})

        random.shuffle(results)
        data_result = get_result_data(results, "test1")

    return data_result

def evaluate_nodes(nodes, user_data, exclusive=False):
    results = {'paths': [], 'outputs': []}

    if not nodes:
        return results

    for node in nodes:
        if not node['enable']:
            continue

        node_result = {'paths': [], 'outputs': []}

        if evaluate_conditions(node['conditions'], user_data, variables):
            node_result['paths'].append(node['description'])

            if node['output']:
                output_result = evaluate_output(node['outputs'], user_data)
                node_result['outputs'].extend(output_result)

            if node['nodes'] and len(node['nodes']) > 0:
                child_result = evaluate_nodes(node['nodes'], user_data, node['exclusive'])
                node_result['paths'].extend(child_result['paths'])
                node_result['outputs'].extend(child_result['outputs'])

            results['paths'].append(node_result['paths'])
            results['outputs'].extend(node_result['outputs'])

            if exclusive:
                break

    return results
